### PR TITLE
Updater DLL replacement & documentation

### PR DESCRIPTION
**Pull request by Starkku**
This PR replaces the updater DLL with a modified one based on the original. Source code for the new DLL and the second-stage updater for now are available in a private repository under CnCNet organization.

Includes a documentation for updater and its features, as well as the [version file writer](https://github.com/Starkku/VersionWriter) that has now been included along with the updater scripts and the second-stage updater executable in the client repository.

Migration information:
- Updater no longer has hardcoded list of download mirrors or custom components. This information must now be set in `UpdaterConfig.ini` (example is included amongst default resources in client repository). For a reference, the previously hardcoded information can be found in format used by `UpdaterConfig.ini` [here](https://gist.github.com/Starkku/1d52f0040d7a00d79e57afc2fba5f97b).
- Second-stage updater no longer has hardcoded list of launcher executables to check for when restarting the client. It will now only check `ClientDefinitions.ini` for `LauncherExe` key, and it it fails to read and launch this the client will not automatically restart after updating.
- Updater DLL filename has been changed from `DTAUpdater.dll` to `ClientUpdater.dll` and second-stage updater from `clientupdt.dat` to `SecondStageUpdater.exe` and has been moved from base folder to `Resources`.

### CODE REVIEW

File: DTAConfig/OptionPanels/ComponentsPanel.cs [hunk 1]
```diff
@@ -204,7 +222,11 @@ private void HandleDownloadProgressChanged(CustomComponent cc, int percentage)
             percentage = Math.Min(percentage, 100);
 
             var btn = installationButtons.Find(b => object.ReferenceEquals(b.Tag, cc));
-            btn.Text = "Downloading..".L10N("UI:DTAConfig:Downloading")+ " "+ percentage + "%";
+
+            if (cc.Archived && percentage == 100)
+                btn.Text = "Unpacking..".L10N("UI:DTAConfig:Unpacking");
+            else
+                btn.Text = "Downloading..".L10N("UI:DTAConfig:Downloading") + " " + percentage + "%";
```
Comment (Metadorius) [inline DTAConfig/OptionPanels/ComponentsPanel.cs:229]: Maybe would be better to have a full 3-dot ellipsis?

File: DTAConfig/OptionPanels/ComponentsPanel.cs [hunk 2]
```diff
@@ -134,30 +151,33 @@ private void Btn_LeftClick(object sender, EventArgs e)
...
-            {
+            {   
+                string archiveSizeMsg = "";
+                if (cc.Archived && cc.RemoteArchiveSize != cc.RemoteSize)
+                    archiveSizeMsg = " (size of the download is " + GetSizeString(cc.RemoteArchiveSize) + ")";     
+                  
                 var msgBox = new XNAMessageBox(WindowManager, "Confirmation Required".L10N("UI:DTAConfig:UpdateConfirmRequiredTitle"),
                     string.Format(("To enable {0} the Client will download the necessary files to your game directory." +
-                    Environment.NewLine + Environment.NewLine + "This will take an additional {1} of disk space, and the download may last" +
+                    Environment.NewLine + Environment.NewLine + "This will take an additional {1} of disk space{2}, and the download may last" +
```
Comment (Metadorius) [inline DTAConfig/OptionPanels/ComponentsPanel.cs:172]: That `{2}` looks a bit wrong, maybe spaces at least are needed? I don't get why it's placed here.

File: DTAConfig/OptionPanels/ComponentsPanel.cs [hunk 3]
```diff
@@ -134,30 +151,33 @@ private void Btn_LeftClick(object sender, EventArgs e)
...
-                Thread thread = new Thread(cc.DownloadComponent);
-                thread.Start();
+                cc.DownloadComponent();
             }
             else
-            {
+            {   
+                string archiveSizeMsg = "";
+                if (cc.Archived && cc.RemoteArchiveSize != cc.RemoteSize)
+                    archiveSizeMsg = " (size of the download is " + GetSizeString(cc.RemoteArchiveSize) + ")";
```
Comment (Metadorius) [inline DTAConfig/OptionPanels/ComponentsPanel.cs:168]: `L10N` call missed.

File: DTAConfig/OptionPanels/ComponentsPanel.cs [hunk 4]
```diff
@@ -134,30 +151,33 @@ private void Btn_LeftClick(object sender, EventArgs e)
                 if (cc.LocalIdentifier == cc.RemoteIdentifier)
                 {
                     File.Delete(ProgramConstants.GamePath + cc.LocalPath);
-                    btn.Text = "Install";
+                    btn.Text = "Install (" + GetSizeString(cc.RemoteSize) + ")";
```
Comment (Metadorius) [inline DTAConfig/OptionPanels/ComponentsPanel.cs:154]: `L10N`?

File: DXMainClient/DXGUI/Generic/MainMenu.cs [hunk 5]
```diff
@@ -632,7 +640,7 @@ private void CheckForUpdates()
             lastUpdateCheckTime = DateTime.Now;
         }
 
-        private void CUpdater_FileIdentifiersUpdated()
+        private void Updater_FileIdentifiersUpdated()
         {
             WindowManager.AddCallback(new Action(HandleFileIdentifierUpdate), null);
         }
```
Comment (Metadorius) [inline DXMainClient/DXGUI/Generic/MainMenu.cs:646]:
```suggestion
        private void Updater_FileIdentifiersUpdated()
            => WindowManager.AddCallback(new Action(HandleFileIdentifierUpdate), null);
```

File: DXMainClient/DXGUI/Generic/ManualUpdateQueryWindow.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            Name = "ManualUpdateQueryWindow";
+            ClientRectangle = new Rectangle(0, 0, 251, 140);
+            BackgroundTexture = AssetLoader.LoadTexture("updatequerybg.png");
+
+            lblDescription = new XNALabel(WindowManager);
+            lblDescription.Name = "lblDescription";
+            lblDescription.ClientRectangle = new Rectangle(12, 9, 0, 0);
+            lblDescription.Text = "Version {0} is available." + Environment.NewLine + Environment.NewLine +
+                "Manual download and installation is" + Environment.NewLine + "required.".L10N("UI:Main:ManualDownloadAvailable");
```
Comment (Metadorius) [inline DXMainClient/DXGUI/Generic/ManualUpdateQueryWindow.cs:36]: Pretty sure you forgot brackets here, it will localize only `required.`

**Review by Metadorius — CHANGES_REQUESTED** [review]
Please check the code for missed L10N calls and expression bodies :P
Otherwise the code is good.

File: DTAConfig/OptionPanels/ComponentsPanel.cs (hunk 2 above)
Comment (Starkku) [reply DTAConfig/OptionPanels/ComponentsPanel.cs:172]: It is replaced by `archiveSizeMsg`, which is conditionally either a proper message or an empty string, hence the lack of space there (the message itself is prefixed with one if it is set.

File: DTAConfig/OptionPanels/ComponentsPanel.cs [hunk 7]
```diff
@@ -74,22 +73,45 @@ public override void Initialize()
...
 
-            int componentIndex = 0;
-            bool buttonEnabled;
+            UpdateInstallationButtons();
+        }
+
+        public override bool Save()
+        {
+            return base.Save();
+        }
```
[the remark this answers is not in the document]
Comment (Starkku) [reply DTAConfig/OptionPanels/ComponentsPanel.cs:95]: Don't think so.

File: ClientCore/ClientConfiguration.cs [hunk 8]
```diff
@@ -200,6 +200,8 @@ public void RefreshSettings()
 
         public string CreditsURL => clientDefinitionsIni.GetStringValue(SETTINGS, "CreditsURL", "http://www.moddb.com/mods/the-dawn-of-the-tiberium-age/tutorials/credits#Rampastring");
 
+         public string ManualDownloadURL => clientDefinitionsIni.GetStringValue(SETTINGS, "ManualDownloadURL", string.Empty);
```
[the remark this answers is not in the document]
Comment (Starkku) [reply ClientCore/ClientConfiguration.cs:203]: Will fix the prefixing whitespaces/tabs but will leave the newlines for later (fix the entire file to look like that at once).

**Comment by Starkku** [comment]
Fixed the issues pointed out.

File: DXMainClient/DXGUI/Generic/MainMenu.cs [hunk 9]
```diff
@@ -538,9 +538,12 @@ public void PostInit()
                     lblUpdateStatus.DrawUnderline = false;
                 }
                 else if (UserINISettings.Instance.CheckForUpdates)
+                {
                     CheckForUpdates();
-                else
+                }
+                {
                     lblUpdateStatus.Text = "Click to check for updates.".L10N("UI:Main:ClickToCheckUpdate");
+                }
```
Comment (Metadorius) [inline DXMainClient/DXGUI/Generic/MainMenu.cs:546]: ?

**Review by Metadorius — CHANGES_REQUESTED** [review]

File: DXMainClient/DXGUI/Generic/MainMenu.cs (hunk 9 above)
Comment (Starkku) [reply DXMainClient/DXGUI/Generic/MainMenu.cs:546]: You said 'inconsistent braces', I added braces to the entire if-elseif-else block.

Comment (Metadorius) [reply DXMainClient/DXGUI/Generic/MainMenu.cs:546]: image

Comment (Starkku) [reply DXMainClient/DXGUI/Generic/MainMenu.cs:546]: Yeah okay must have accidentally wiped the `else` out. Weird that it generates no build errors.

Comment (Metadorius) [reply DXMainClient/DXGUI/Generic/MainMenu.cs:546]: Probably because operator braces can be used without anything else both in C# and C++.

**Pull request merged**